### PR TITLE
adding annotation-service and annotation-revision to kn service create/update

### LINKED ISSUE #1014: Add "--annotation-service" and "--annotation-revision" to kn service create/update
### Feature request

In order to provide the same support as for labels also for annotations, the two options mentioned in the title should be implemented with the same semantics as for labels:

```
-l, --label stringArray             Labels to set for both Service and Revision. name=value; you may provide this flag any number of times to set
                                      multiple labels. To unset, specify the label name followed by a "-" (e.g., name-).
      --label-revision stringArray    Revision label to set. name=value; you may provide this flag any number of times to set multiple labels. To unset,
                                      specify the label name followed by a "-" (e.g., name-). This flag takes precedence over "label" flag.
      --label-service stringArray     Service label to set. name=value; you may provide this flag any number of times to set multiple labels. To unset,
                                      specify the label name followed by a "-" (e.g., name-). This flag takes precedence over "label" flag.
```
### Use case

One use case is to add `sidecar.istio.io/inject: "true"` annotation to the pod template which is currently not possible. 

cc: @nak3

### PR DESCRIPTION
**Pull request by arghya88**
## Description

Add "--annotation-service" and "--annotation-revision" flag to to kn service create and update providing ability to set specific annotations on service and revision 

## Changes

* Add "--annotation-service" and "--annotation-revision" to kn service create/update

## Reference

Fixes #1014

/lint

### CODE REVIEW

**Comment by dsimansk** [comment]
Thanks looks good @arghya88 !

Could you please take a look on unit and e2e tests to add coverage for those specific flags?

/ok-to-test
/cc @navidshaikh

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 1]
```diff
@@ -406,25 +422,46 @@ func (p *ConfigurationEditFlags) Apply(
 		}
 	}
 
-	if cmd.Flags().Changed("annotation") {
-		annotationsMap, err := util.MapFromArrayAllowingSingles(p.Annotations, "=")
+	if cmd.Flags().Changed("annotation") || cmd.Flags().Changed("annotation-service") || cmd.Flags().Changed("annotation-revision") {
```
Comment (dsimansk) [inline pkg/kn/commands/service/configuration_edit_flags.go:425]: Maybe, I'd prefer separate processing per changed flag, but let's wait for other opinions.

/cc @maximilien @rhuss @navidshaikh

Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go:425]: I have followed the pattern used for "label-service" and label-revision" for better consistency

**Review by navidshaikh — CHANGES_REQUESTED** [review]

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 2]
```diff
@@ -154,6 +156,20 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
 			"To unset, specify the annotation name followed by a \"-\" (e.g., name-).")
 	p.markFlagMakesRevision("annotation")
 
+	command.Flags().StringArrayVarP(&p.AnnotationsService, "annotation-service", "", []string{},
+		"Service annotation to set. name=value; you may provide this flag "+
+			"any number of times to set multiple annotation. "+
```
[the remark this answers is not in the document]
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Changes done for this in latest commit

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 3]
```diff
@@ -154,6 +156,20 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
...
+			"To unset, specify the annotation name followed by a \"-\" (e.g., name-). This flag takes "+
+			"precedence over \"annotation\" flag.")
+	p.markFlagMakesRevision("annotation-service")
+
+	command.Flags().StringArrayVarP(&p.AnnotationsRevision, "annotation-revision", "", []string{},
+		"Revision annotation to set. name=value; you may provide this flag "+
+			"any number of times to set multiple labels. "+
```
[the remark this answers is not in the document]
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Changes done for this in latest commit

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 4]
```diff
@@ -406,25 +422,46 @@ func (p *ConfigurationEditFlags) Apply(
...
 				serviceAnnotations[key] = value
 			}
 		}
-		// Add all user provided annotations to RevisionTemplate
-		err = servinglib.UpdateRevisionTemplateAnnotations(template, annotationsMap, annotationsToRemove)
-		if err != nil {
-			return err
+
+		for key, value := range annotationServiceFlagMap {
+			if !strings.HasPrefix(key, autoscaling.GroupName) {
```
[the remark this answers is not in the document]
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go]: Changes done for this in latest commit

[the remark this answers is not in the document]
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go]: So we fail the service creation if there is a autoscaling annotation in service annotations?

File: docs/cmd/kn_service_create.md [hunk 5]
```diff
@@ -53,48 +53,50 @@ kn service create NAME --image IMAGE
...
-      --scale-min int                 Minimum number of replicas.
-      --service-account string        Service account name to set. An empty argument ("") clears the service account. The referenced service account must exist in the service's namespace.
-      --user int                      The user ID to run the container (e.g., 1001).
-      --volume stringArray            Add a volume from a ConfigMap (prefix cm: or config-map:) or a Secret (prefix secret: or sc:). Example: --volume myvolume=cm:myconfigmap or --volume myvolume=secret:mysecret. You can use this flag multiple times. To unset a ConfigMap/Secret reference, append "-" to the name, e.g. --volume myvolume-.
-      --wait                          Wait for 'service create' operation to be completed. (default true)
-      --wait-timeout int              Seconds to wait before giving up on waiting for service to be ready. (default 600)
+  -a, --annotation stringArray            Service annotation to set. name=value; you may provide this flag any number of times to set multiple annotations. To unset, specify the annotation name followed by a "-" (e.g., name-).
```
Comment (rhuss) [inline docs/cmd/kn_service_create.md:56]: You should mention that this annotation is on Service *and* the Revision. Same like for label.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 6]
```diff
@@ -154,6 +156,20 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
 			"To unset, specify the annotation name followed by a \"-\" (e.g., name-).")
 	p.markFlagMakesRevision("annotation")
 
+	command.Flags().StringArrayVarP(&p.AnnotationsService, "annotation-service", "", []string{},
+		"Service annotation to set. name=value; you may provide this flag "+
+			"any number of times to set multiple annotations. "+
+			"To unset, specify the annotation name followed by a \"-\" (e.g., name-). This flag takes "+
+			"precedence over \"annotation\" flag.")
```
Comment (rhuss) [inline pkg/kn/commands/service/configuration_edit_flags.go:163]:
```suggestion
			"precedence over the \"annotation\" flag.")
```

Comment (rhuss) [reply pkg/kn/commands/service/configuration_edit_flags.go:163]: If you could add the missing article also for the labels doc, that would be awesome.

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 7]
```diff
@@ -154,6 +156,20 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
...
+	p.markFlagMakesRevision("annotation-service")
+
+	command.Flags().StringArrayVarP(&p.AnnotationsRevision, "annotation-revision", "", []string{},
+		"Revision annotation to set. name=value; you may provide this flag "+
+			"any number of times to set multiple annotations. "+
+			"To unset, specify the annotation name followed by a \"-\" (e.g., name-). This flag takes "+
+			"precedence over \"annotation\" flag.")
```
Comment (rhuss) [inline pkg/kn/commands/service/configuration_edit_flags.go:170]:
```suggestion
			"precedence over the \"annotation\" flag.")
``

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 1 above)
Comment (rhuss) [reply pkg/kn/commands/service/configuration_edit_flags.go:425]: I'm fine with that (we can optimize later if we feel the need)

**Review by rhuss — COMMENTED** [review]
thanks, looks good to me (with some minor comments).

Please could you also add enty to the CHANGELOG.adoc ?

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 8]
```diff
@@ -149,11 +151,25 @@ func (p *ConfigurationEditFlags) addSharedFlags(command *cobra.Command) {
...
+			"any number of times to set multiple annotations. "+
+			"To unset, specify the annotation name followed by a \"-\" (e.g., name-). This flag takes "+
+			"precedence over the \"annotation\" flag.")
+	p.markFlagMakesRevision("annotation-service")
+
+	command.Flags().StringArrayVarP(&p.AnnotationsRevision, "annotation-revision", "", []string{},
+		"Revision annotation to set. name=value; you may provide this flag "+
+			"any number of times to set multiple annotations. "+
+			"To unset, specify the annotation name followed by a \"-\" (e.g., name-). This flag takes "+
+			"precedence over the \"annotation\" flag.")
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:170]: I can comment out lines 171 and/or 164 and all tests pass!

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 9]
```diff
@@ -406,25 +422,47 @@ func (p *ConfigurationEditFlags) Apply(
 		}
 	}
 
-	if cmd.Flags().Changed("annotation") {
-		annotationsMap, err := util.MapFromArrayAllowingSingles(p.Annotations, "=")
+	if cmd.Flags().Changed("annotation") || cmd.Flags().Changed("annotation-service") || cmd.Flags().Changed("annotation-revision") {
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:425]: I can comment out the last condition in this if statement to read:

```go

if cmd.Flags().Changed("annotation") || cmd.Flags().Changed("annotation-service") { //|| cmd.Flags().Changed("annotation-revision") {	
...
}
```

Yet all tests are passing...

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 10]
```diff
@@ -406,25 +422,47 @@ func (p *ConfigurationEditFlags) Apply(
...
+		annotationsAllMap, err := util.MapFromArrayAllowingSingles(p.Annotations, "=")
 		if err != nil {
 			return fmt.Errorf("Invalid --annotation: %w", err)
 		}
+		annotationRevisionFlagMap, err := util.MapFromArrayAllowingSingles(p.AnnotationsRevision, "=")
+		if err != nil {
```
Comment (maximilien) [inline pkg/kn/commands/service/configuration_edit_flags.go:431]: I can comment out lines 431-433 and all tests pass

**Review by maximilien — CHANGES_REQUESTED** [review]
Thanks for contributions. If possible try to address lack of coverage in unit tests and add an integration test. Thanks.

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 10 above)
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go:431]: Have refactored the tests

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 9 above)
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go:425]: Have refactored the tests

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 8 above)
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go:170]: Have refactored the tests

File: test/e2e/service_options_test.go [hunk 11]
```diff
@@ -163,6 +163,21 @@ func TestServiceOptions(t *testing.T) {
 	validateServiceInitScale(r, "svc11", "2")
 	t.Log("delete service")
 	test.ServiceDelete(r, "svc11")
+
```
Comment (rhuss) [inline test/e2e/service_options_test.go:166]: From the tests, what I'm missing a bit are the following cases:

* `--annotation` set the same annotation on *both*, service and revision
* `--annotation-revision` alone does *not* set the service annotation (the check is only that the revision annotation is set)
* Same negative check for service annotations is missing.

Also I would refactor the test to have a single validation method, that takes two maps for service and revision annotations respectively where each of the maps can be also nil it should not be set.

Comment (rhuss) [reply test/e2e/service_options_test.go:166]: otherwise, looks good to me so we are close to get the PR merged.

**Review by rhuss — COMMENTED** [review]
Thanks !
/approve
/lgtm

**Comment by maximilien** [comment]
/test pull-knative-client-integration-tests

@arghya88 thanks for addressing feedback. Cheers 🍻

File: test/e2e/service_options_test.go (hunk 11 above)
Comment (arghya88) [reply test/e2e/service_options_test.go:166]: I tried the negative tests with `len(annotation) == 0` and `annotation == nil`. Both fails. I think some annotations always get set automatically by default. So how do we test that no service annotations is set when we use --annotation-revision ?

**Comment by navidshaikh** [comment]
this test needs fix [log:ref](https://prow.knative.dev/view/gs/knative-prow/pr-logs/pull/knative_client/1029/pull-knative-client-integration-tests/1311190880333336576#1:build-log.txt%3A907)
```
    service_options_test.go:167: create and validate service and revision annotations
    service_options_test.go:278: assertion failed: 
        --- gotRevisionAnnotations
        +++ expectedRevisionAnnotations
          map[string]string{
        - 	"client.knative.dev/user-image": "gcr.io/knative-boskos-70/client-e2e-img/16051/helloworld:latest",
          	"rev":                           "helloworld-rev",
          }        
```

**Comment by navidshaikh** [comment]
/retest


        result_collector.go:75: ERROR: Error: ReconcileIngressFailed: Ingress reconciliation failed

**Comment by navidshaikh** [comment]
/retest

phew...
```
ERROR: timeout waiting for pods to come up
```

File: pkg/kn/commands/service/configuration_edit_flags.go [hunk 12]
```diff
@@ -406,25 +422,47 @@ func (p *ConfigurationEditFlags) Apply(
...
 			}
 		}
-		// Add all user provided annotations to RevisionTemplate
-		err = servinglib.UpdateRevisionTemplateAnnotations(template, annotationsMap, annotationsToRemove)
-		if err != nil {
-			return err
+
+		for key, value := range annotationServiceFlagMap {
+			if strings.HasPrefix(key, autoscaling.GroupName) {
+				return fmt.Errorf("Service cannot have annotation: %s", key)
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply pkg/kn/commands/service/configuration_edit_flags.go:461]: can be done in a subsequent PR

File: pkg/kn/commands/service/create_mock_test.go [hunk 13]
```diff
@@ -579,6 +579,106 @@ func TestServiceCreateWithAnnotations(t *testing.T) {
...
+func TestServiceCreateWithAutoScaleServiceAnnotationsError(t *testing.T) {
+	client := knclient.NewMockKnServiceClient(t)
+
+	r := client.Recorder()
+
+	output, err := executeServiceCommand(client, "create", "foo", "--image", "gcr.io/foo/bar:baz",
+		"--annotation-service", autoscaling.InitialScaleAnnotationKey+"=1",
+		"--no-wait", "--revision-name=")
+	assert.Assert(t, err != nil)
+	assert.Assert(t, util.ContainsAll(output, "Service cannot have annotation: autoscaling.knative.dev/initialScale"))
```
Comment (navidshaikh) [inline pkg/kn/commands/service/create_mock_test.go:677]: the error message would need update following aforementioned suggestion

(also `ContainsAll` can work with array of words, so we should check presence of the words in the error message than the complete sentence)

**Review by navidshaikh — APPROVED** [review]
/lgtm

added a suggestion to update error message, which can be done in a subsequent PR

File: pkg/kn/commands/service/configuration_edit_flags.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (arghya88) [reply pkg/kn/commands/service/configuration_edit_flags.go:461]: Sure...I will create a new follow up PR to address this.

File: pkg/kn/commands/service/create_mock_test.go (hunk 13 above)
Comment (arghya88) [reply pkg/kn/commands/service/create_mock_test.go:677]: Will take care of this in the new follow up PR.

**Comment by navidshaikh** [comment]
/retest


      result_collector.go:75: ERROR: Error: ReconcileIngressFailed: Ingress reconciliation failed

**Pull request merged**